File: Problem-94-Bin_tree_inorder_traversal/my_sol.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        vector<int> output;
        if (!root) return output;  // base case

        vector<int> leftPart = inorderTraversal(root->left);
        output.insert(output.end(), leftPart.begin(), leftPart.end());

        output.push_back(root->val);

        vector<int> rightPart = inorderTraversal(root->right);
        output.insert(output.end(), rightPart.begin(), rightPart.end());

        return output;
    }
};


// Time Complexity: O(n)

// Space Complexity: O(n)

File: Problem-238-Product of Array Except Self/my_sol.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        size = len(nums)

        left = [1] * size
        for i in range(1, size):
                left[i] = left[i-1] * nums[i-1]

        right = [1] * size
        for i in range(size - 2, -1, -1):
                right[i] = right[i+1] * nums[i+1]


        output = [1] * size 
        for i in range(size):
            output =  left[i] * right[i]
        
        return output


# Time Complexity: O(N)
# Space Complexity: O(N)